File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/model/QueryContext.java
package org.cqframework.cql.cql2elm.model;

import org.hl7.cql.model.DataType;
import org.hl7.cql.model.ListType;
import org.hl7.elm.r1.AliasedQuerySource;
import org.hl7.elm.r1.LetClause;

import java.util.Collection;
import java.util.HashMap;

public class QueryContext {
    private final HashMap<String, AliasedQuerySource> sources = new HashMap<>();
    private final HashMap<String, LetClause> lets = new HashMap<>();

    private void internalAddQuerySource(AliasedQuerySource source) {
        sources.put(source.getAlias(), source);
    }

    // Adds a related (i.e. with or without) source, which does not change cardinality of the query
    public void addRelatedQuerySource(AliasedQuerySource source) {
        internalAddQuerySource(source);
    }

    // Adds primary sources, which affect cardinality (any primary plural source results in a plural query)
    public void addPrimaryQuerySources(Collection<AliasedQuerySource> sources) {
        for (AliasedQuerySource source : sources) {
            internalAddQuerySource(source);
            if (source.getResultType() instanceof ListType) {
                isSingularValue = false;
            }
        }
    }

    public Collection<AliasedQuerySource> getQuerySources() {
        return sources.values();
    }

    public void removeQuerySource(AliasedQuerySource source) {
        sources.remove(source.getAlias());
    }

    public void removeQuerySources(Collection<AliasedQuerySource> sources) {
        for (AliasedQuerySource source : sources) {
            removeQuerySource(source);
        }
    }

    public void addLetClauses(Collection<LetClause> lets) {
        for (LetClause let : lets) {
            addLetClause(let);
        }
    }

    public void addLetClause(LetClause let) {
        lets.put(let.getIdentifier(), let);
    }

    public void removeLetClause(LetClause let) {
        lets.remove(let.getIdentifier());
    }

    public void removeLetClauses(Collection<LetClause> lets) {
        for (LetClause let : lets) {
            removeLetClause(let);
        }
    }

    public AliasedQuerySource resolveAlias(String identifier) {
        return sources.get(identifier);
    }

    public ResolvedIdentifierList resolveCaseIgnoredAliases(String identifier) {
        final ResolvedIdentifierList resolvedIdentifierList = new ResolvedIdentifierList();

        sources.entrySet()
                .stream()
                .filter(k -> k.getKey().equalsIgnoreCase(identifier) && !k.getKey().equals(identifier))
                .forEach(entry -> resolvedIdentifierList.addResolvedIdentifier(new ResolvedIdentifier(entry.getKey(), MatchType.CASE_IGNORED, entry.getValue())));

        return resolvedIdentifierList;
    }

    public LetClause resolveLet(String identifier) {
        return lets.get(identifier);
    }

    public ResolvedIdentifierList resolveCaseIgnoredLets(String identifier) {
        final ResolvedIdentifierList resolvedIdentifierList = new ResolvedIdentifierList();

        lets.entrySet()
                .stream()
                .filter(k -> k.getKey().equalsIgnoreCase(identifier) && !k.getKey().equals(identifier))
                .forEach(entry -> resolvedIdentifierList.addResolvedIdentifier(new ResolvedIdentifier(entry.getKey(), MatchType.CASE_IGNORED, entry.getValue())));

        return resolvedIdentifierList;
    }

    private boolean isSingularValue = true;
    public boolean isSingular() {
        return isSingularValue;
    }

    private boolean inSourceClauseValue;
    public void enterSourceClause() {
        inSourceClauseValue = true;
    }

    public void exitSourceClause() {
        inSourceClauseValue = false;
    }

    public boolean inSourceClause() {
        return inSourceClauseValue;
    }

    private boolean inSortClauseValue;
    public void enterSortClause() {
        inSortClauseValue = true;
    }

    public void exitSortClause() {
        inSortClauseValue = false;
    }

    public boolean inSortClause() {
        return inSortClauseValue;
    }

    private boolean isImplicitValue;
    public boolean isImplicit() { return isImplicitValue; }
    public void setIsImplicit(boolean isImplicitValue) {
        this.isImplicitValue = isImplicitValue;
    }

    private DataType resultElementType;
    public DataType getResultElementType() {
        return resultElementType;
    }

    public void setResultElementType(DataType resultElementType) {
        this.resultElementType = resultElementType;
    }

    private boolean referencesSpecificContextValue;
    public boolean referencesSpecificContext() {
        return referencesSpecificContextValue;
    }

    public void referenceSpecificContext() {
        referencesSpecificContextValue = true;
    }
}

File: Src/java/cql-to-elm/src/commonMain/kotlin/org/cqframework/cql/cql2elm/model/QueryContext.kt
package org.cqframework.cql.cql2elm.model

import org.cqframework.cql.cql2elm.tracking.Trackable.resultType
import org.hl7.cql.model.DataType
import org.hl7.cql.model.ListType
import org.hl7.elm.r1.AliasedQuerySource
import org.hl7.elm.r1.LetClause

@Suppress("TooManyFunctions")
class QueryContext {
    private val sources = HashMap<String, AliasedQuerySource>()
    private val lets = HashMap<String, LetClause>()

    private fun internalAddQuerySource(source: AliasedQuerySource) {
        sources[source.alias!!] = source
    }

    // Adds a related (i.e. with or without) source, which does not change cardinality of the query
    fun addRelatedQuerySource(source: AliasedQuerySource) {
        internalAddQuerySource(source)
    }

    // Adds primary sources, which affect cardinality (any primary plural source results in a plural
    // query)
    fun addPrimaryQuerySources(sources: Collection<AliasedQuerySource>) {
        for (source in sources) {
            internalAddQuerySource(source)
            if (source.resultType is ListType) {
                isSingular = false
            }
        }
    }

    val querySources: Collection<AliasedQuerySource>
        get() = sources.values

    fun removeQuerySource(source: AliasedQuerySource) {
        sources.remove(source.alias)
    }

    fun removeQuerySources(sources: Collection<AliasedQuerySource>) {
        for (source in sources) {
            removeQuerySource(source)
        }
    }

    fun addLetClause(let: LetClause) {
        lets[let.identifier!!] = let
    }

    private fun removeLetClause(let: LetClause) {
        lets.remove(let.identifier)
    }

    fun removeLetClauses(lets: Collection<LetClause>) {
        for (let in lets) {
            removeLetClause(let)
        }
    }

    fun resolveAlias(identifier: String): AliasedQuerySource? {
        return sources[identifier]
    }

    fun resolveLet(identifier: String): LetClause? {
        return lets[identifier]
    }

    var isSingular: Boolean = true
        private set

    private var inSourceClauseValue = false

    fun enterSourceClause() {
        inSourceClauseValue = true
    }

    fun exitSourceClause() {
        inSourceClauseValue = false
    }

    fun inSourceClause(): Boolean {
        return inSourceClauseValue
    }

    private var inSortClauseValue = false

    fun enterSortClause() {
        inSortClauseValue = true
    }

    fun exitSortClause() {
        inSortClauseValue = false
    }

    fun inSortClause(): Boolean {
        return inSortClauseValue
    }

    var isImplicit: Boolean = false
    var resultElementType: DataType? = null
    var referencesSpecificContextValue = false
}
